### ISSUE TITLE
Enumeration type support in Structorizer

### ISSUE THREAD

**Issue by codemanyak** [fesch/Structorizer.Desktop#739]
A possible syntax for an enumeration type definition in Structorizer might be:
`type cb_support = enum{CB_OK, CB_WARNING, CB_ARCHAIC, CB_OBSOLETE <- 9, CB_SKIP, CB_IGNORE, CB_ERROR, CB_UNCONFORMABLE}`
The effect for Analyser and Executor would be the same as from the provisional const definitions, but on code export there would be the chance to produce veritable enumeration types.

Note that an enumerator type definition of this kind would have an actual impact on execution (i.e. defining all comprised constants at once) in contrast to the record type definitions. It is of course a problem that we introduce more and more non-algorithmic stuff into a structogram, which isn't intended to contain non-algorithmic elements. I do see that the pragmatic enhancements to support code export and import are compromising the original ideas of Nassi and Shneiderman.
It might be a good idea to return to an idea privately proposed by Klaus-Peter Reimers to provide a separate assistant area in Structorizer for data type construction. Of course a clever selective connection between algorithmic diagrams and the type pool would still have to be invented.

_Originally posted by @codemanyak in https://github.com/fesch/Structorizer.Desktop/issues/558#issuecomment-405122276_

**Comment by codemanyak** [comment]
The implementation is advancing. There is still a slight syntactic modification (symbol `=` instead of `<-` for enumerator code assignments) might be changed back.
![enum_demo_snippet_3 30-03](https://user-images.githubusercontent.com/15326471/69441013-281dbd00-0d4a-11ea-8f8a-206f08de6d09.png)

Variable display in the control panel will show the type name with the generated constants and tries to show the symbolic names rather than the codes for declared variables (providing a choice list on editing):
![enum_demo_exec3_3 30-03](https://user-images.githubusercontent.com/15326471/69441047-379d0600-0d4a-11ea-8e40-4a8694fc1b1b.png)
Type inference does not work very deeply, though, and may fail.

**Comment by GitMensch** [comment]
Looks nice. @codemanyak What is the status of the exporters (C, C++, Java, Perl, COBOL) for this?

**Comment by codemanyak** [comment]
Well, I did my very best.
Just the COBOL generator project has not advanced yet.

... and I just had to learn that the Perl generator does not cope with an initializing variable declaration as in:
```
type Drink = enum{COFFEE, TEA, BEER, WATER}
var drink: Drink <- TEA
```

**Comment by codemanyak** [comment]
By the way, how would an enumeration type sensibly be translated to COBOL? You would have to define the associated value names as pic 9(n) constants, wouldn't you? At least I haven't found anything like an enumeration type in the COBOL documentation.

**Comment by GitMensch** [comment]
> By the way, how would an enumeration type sensibly be translated to COBOL? You would have to define the associated value names as pic 9(n) constants, wouldn't you? At least I haven't found anything like an enumeration type in the COBOL documentation.

This depends on how you use this

* Option 1: "enum that may be used multiple times" - constants (common COBOL85 extension "level 78" constant or COBOL2002+ constant):

```cobol
78 coffee VALUE 1.
78 tea    VALUE 2.
78 beer   VALUE 3.
78 water  VALUE 9.
77 drink1 PIC 9 VALUE coffee.
77 drink2 PIC 9.
...
IF drink1 = tea
   MOVE beer TO drink2
END-IF
```

* Option 2: an "enum" that is only used once (COBOL74 or earlier [I'm too young ;-] +)
```cobol
01 drink  PIC 9 VALUE 2.
88 coffee VALUE 1.
88 tea    VALUE 2.
88 beer   VALUE 3.
88 water  VALUE 9.
...
IF coffee
   SET tea TO TRUE
ELSE
  MOVE 0 TO drink
END-IF
```


* Option 3: an "enum" as typedef that is used multiple times (COBOL 2002+)
```cobol
01 drink  PIC 9 IS TYPEDEF.
88 coffee VALUE 1.
88 tea    VALUE 2.
88 beer   VALUE 3.
88 water  VALUE 9.

01 my-drink   TYPE drink VALUE 1.
01 your-drink TYPE drink.
...
*> COBOL beauty, we can use the better sounding IN instead of OF here :-)
IF coffee IN my-drink
   SET tea IN your-drink TO TRUE
ELSE
  MOVE my-drink TO your-drink
END-IF
```

only the first one would allow a "named inline initialization", for all others either a separate `SET` statement or (in real world) a `VALUE` clause using the appropriate number would be used.

**Comment by codemanyak** [comment]
@GitMensch 
Thank you! In my opinion indeed the 78-level constants are what makes most sense here. The 88-level conditions are not actually compatible with the way enum values are used in other languages and Structorizer.

**Comment by GitMensch** [comment]
@codemanyak just asking for a pingback - are the level 78 constants used in the enumeration support already?

**Comment by codemanyak** [comment]
@GitMensch 
> are the level 78 constants used in the enumeration support already?

No, since I haven't found any hint in the COBOL documentation that some kind of relation is detectable between several level-78 constants. How shall the parser find out whether some of them are supposed to form a common enumerable domain?

**Comment by GitMensch** [comment]
> No, since I haven't found any hint in the COBOL documentation that some kind of relation is detectable between several level-78 constants. How shall the parser find out whether some of them are supposed to form a common enumerable domain?

Not at all, there is no reasonable option to tweak the COBOL import here.

> Or are we talking about the - still unimplemented - COBOL generator? I prepared it in the generator fragment.

yes :-) so no need for an issue other than referencing the issue the COBOL generator was tracked under.

**Comment by codemanyak** [comment]
The enum type support had been forgotten for C# export up to and including version 3.32-11.

**Comment by codemanyak** [comment]
C# export implemented with version 3.32-12.
